### PR TITLE
test/static-code: introduce ruff and fix vulture

### PR DESCRIPTION
**Pull request by allisonkarlitskaya**

### CODE REVIEW

File: src/cockpit/samples.py
Comment (martinpitt) [inline src/cockpit/samples.py]: This should then be consistent and use `_` everywhere. Moreover, it then seems nicer to directly pick the 4 values that we care about from the array positions, instead of having this soup of `_` and many wasted lines.

**Review by martinpitt — CHANGES_REQUESTED** [review]
Whoa, that was quite a rabbit hole!

File: tools/vulture_whitelist/cockpit.py
Comment (allisonkarlitskaya) [reply tools/vulture_whitelist/cockpit.py:1]: I generated it by hand, although there's also `vulture --make-whitelist`, although it produces some pretty terrible output.

According to the docs https://github.com/jendrikseipp/vulture/blob/main/README.md the preferred method for marking stuff as used is by using whitelist files.  At first I shared your opinion but after a day of working with it, I started to agree.

I have to admit that for everything in the `cockpit.py` file, we should try to eliminate this.  I have an idea for a tweak to the `systemd_ctypes` API that would help with that, for example.

For the stuff in other libraries, I really like this form of indicating use.

And particularly for the "our stuff that's used in external git repositories" for `testlib.py` I'm really quite fond of this syntax.  It basically apes as if we did use it ourselves.  It's very intuitive.

There were a few cases where the whitelist files didn't feel right — mostly for stuff in tests which wasn't a part of any public API (our API, standard library, or other).  For those couple of cases I did use `# noqa` since it felt like the better option.

But taking the case of `Browser.get_checked()` as an example: writing some `# noqa this is used in cockpit-machines yada yada` seems much worse than containing that information in a separate file.

As mentioned in the docs, linked above:

> However, we recommend using whitelists instead of noqa comments, since noqa comments add visual noise to the code and make it harder to read.

... and after a day of playing around with this stuff, I pretty much agree.

File: test/static-code
Comment (allisonkarlitskaya) [reply test/static-code:14]: I was kinda wondering why it worked at all since it contained an outright error.  I considered dropping it as well to see if it even broke anything.  Happy to see this go.

File: test/common/lcov.py
Comment (allisonkarlitskaya) [reply test/common/lcov.py:112]: Sure.  I did that to document that we expect there to be a third parameter here as well, and to assert that it exists.  I could just as well do that with a comment, or skip it entirely...

File: test/common/vulture_whitelist.py
Comment (allisonkarlitskaya) [reply test/common/vulture_whitelist.py:3]: I don't agree this is ugly.  See my reply to your other note, below.

File: test/common/testlib.py
Comment (allisonkarlitskaya) [reply test/common/testlib.py:2171]: The "standard practice" here was maybe something I got confused about.  I thought I had fixed a `mypy` issue in `systemd_ctypes` by doing this after reading something online, but I can't find either of those things anymore, so maybe I was imagining something else.

The error that we fix by this is that `vulture` finds no user of this variable that we set.  The reason for that is that we access it like so:

```python
  def get_decorator(method, _class, name, default=None):
      """Get decorator value of a test method or its class
  
      Return None if the decorator was not set.
      """
      attr = "_testlib__" + name
      return getattr(method, attr, getattr(_class, attr, default))
```

So this is basically about using `setattr()` to match `getattr()`.

Maybe we want a helper like `set_decorator()` or so to do with it, but this would end up being very little more than a single `setattr()` call...

If nothing else, this commit could probably use a better commit message.

File: test/pytest/test_packages.py
Comment (allisonkarlitskaya) [reply test/pytest/test_packages.py:115]: Oops.  I did remove `add_conditional_manifest()` and even eluded to it in the commit log:

>  This let's us drop our existing helper which was specific to testing conditionals since the new one is sufficient.

but the fragment that actually did that ended up in the wrong commit.  I'll rebase that.

File: pyproject.toml
Comment (allisonkarlitskaya) [reply pyproject.toml]: This happens in a later commit.

The idea was to add ruff, and then firm things up as we fix them.  See the commit message:

>  The initial configuration is chosen to avoid having errors straight away.  Fixes will follow.

File: containers/unit-tests/setup.sh
Comment (allisonkarlitskaya) [reply containers/unit-tests/setup.sh:67]: Ya.  Nice argument name, eh? :)

It's basically a way of saying, ya, we're doing an install into a world that's supposed to be handled by the system package manager.  If we don't do that, we get a big message about "use a venv, dummy!!"

Comment (martinpitt) [reply containers/unit-tests/setup.sh:67]: You mean that installs into /usr, not /usr/local? I cannot confirm that. In a cockpit/tasks toolbox, I ran `sudo pip install ruff`, which installs /usr/local/bin/ruff. Sure I get that "Running pip as the 'root' user" blabla warning, but :man_shrugging: Also, this option sounds both scary and doesn't even appear in `pip install --help`.

File: test/pytest/test_code.py
Comment (martinpitt) [reply test/pytest/test_code.py:36]: No no, I was just curious.

File: test/common/testlib.py
Comment (martinpitt) [reply test/common/testlib.py:2171]: Aah, now I at least understand where this is coming from. Explaining it in the commit message and keeping as-is WFM. Thanks!

File: tools/vulture_whitelist/cockpit.py
Comment (martinpitt) [reply tools/vulture_whitelist/cockpit.py:1]: I still heavily disagree with this. testlib.py is external API, so some unused code needs to be tolerated. But if vulture finds dead code in src/cockpit/, then these are either a lot of false positives (in which case the test isn't very good), or it's actually dead code, in which case that whitelist failure just papers over it, and will prevent us from ever cleaning it up. `# noqa` comments of course don't really change that, but they at least make the vulture overrides explicit and keep that information at one place.

If you feel strongly that this is the right thing to do, I won't block the PR because of this, but I find this an unmaintainable and brittle hack.. (Sorry for the strong words, but I really feel that)

File: test/common/vulture_whitelist.py
Comment (martinpitt) [reply test/common/vulture_whitelist.py:3]: Resolving this thread, discussing in the other one.

File: tools/vulture_whitelist/cockpit.py
Comment (allisonkarlitskaya) [reply tools/vulture_whitelist/cockpit.py:1]: So looking at the whitelist files, we have a few different categories:

 - standard library stuff that, arguably, should go upstream.  They have whitelist files for the standard library and are open to improvements.
 - two oddities from Gtk that probably will never belong properly anywhere, but could also possibly be replaced by using accessor functions instead of attributes
 - a handful of D-Bus properties on various objects which could be avoided by better `systemd_ctypes` API
 - external users of testlib API, which... well, we need some way of documenting "we want to keep this because other people are using it"...

We can iterate a bit on the above ideas, but eventually some things will have to stay.

I guess my challenge to you would be to come up with a better way of dealing with this stuff...

Comment (martinpitt) [reply tools/vulture_whitelist/cockpit.py:1]: The stdlib and Gtk bits are the ones I don't understand at all -- these are external libraries, why is there even a complaint about these? Shouldn't vulture only look at code which is defined in the given source directory? Enumerating random parts of the standard library seems entirely pointless to me. If it doesn't understand the source directory properly, then perhaps it could be called with a bunch of `--ignore-names` for the standard library classes, together with a pointer to an upstream bug report, as this is really just broken. Hopefully these can go away over time with a fixed vulture.

> I guess my challenge to you would be to come up with a better way of dealing with this stuff...

As I already said, I'd prefer some kind of `# noqa external-api` annotation for the testlib and cockpit.User classes. vulture can ignore a decorator, if that can be applied to whole classes, that'd be fine for me. Alternatively, `--ignore-names cockpit.User` etc. would actually be fine for this as well, with an "external API comment".

File: pyproject.toml
Comment (jelly) [inline pyproject.toml:24]: Reading https://lwn.net/Articles/930487/ and adding ruff to my own Python projects, I was wondering. Should we in the future extend the selected rules? I noticed ruff has a **ton** of rules!

File: containers/unit-tests/setup.sh
Comment (allisonkarlitskaya) [reply containers/unit-tests/setup.sh:67]: It's a Debianism.

File: test/common/testlib.py
Comment (allisonkarlitskaya) [reply test/common/testlib.py:283]: Restored.

File: test/common/testlib.py
Comment (allisonkarlitskaya) [reply test/common/testlib.py:626]: Restored.

File: containers/unit-tests/start
Comment (allisonkarlitskaya) [reply containers/unit-tests/start:56]: This is ignoring two things:

 - `_rmtree` doesn't seem to be recognised by vulture as something that the standard library calls.  We could work around that by putting it in a whitelist file, but:
 - `ignore_errors` is also not recognised as being used.

I'm not sure what to do about this case...

File: test/common/lcov.py
Comment (allisonkarlitskaya) [reply test/common/lcov.py:112]: :heavy_check_mark:

File: test/static-code
Comment (allisonkarlitskaya) [reply test/static-code:14]: Dropped this again, thanks.

File: pyproject.toml
Comment (allisonkarlitskaya) [reply pyproject.toml:24]: The idea of this PR was to get things online without too many changes :sweat_smile:

**Comment by martinpitt** [comment]
@allisonkarlitskaya : I re-reviewed, and it seems "test an idea" worked out -- [f37](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18750-20230510-101157-304f2a02-fedora-37/log.html) is past the ABRT tests. :tada: This will address the remaining two outstanding review comments, and I didn't find any new ones. So please squash/message this properly, and we are good to go!

**Review by martinpitt — APPROVED** [review]
Thanks! Assuming green, this looks good now. We need to remember to rebuild the unit-tests container right after landing.

**Comment by martinpitt** [comment]
@allisonkarlitskaya : There was one follow-up for me: mock-bugzilla-server.py is being used by `TestJournal.testAbrtReport{,Cancel}`. Not starting it or even changing the mock version 42 to 43 fails the tests. However, it just uses the `Bugzilla` class/interface, not the two others you removed. So I think it's all good now!

**Pull request merged**
